contracts/src/token/erc20/extensions/flash_mint.rs

```diff
@@ -0,0 +1,447 @@
+            return Err(Error::UnsupportedToken(ERC3156UnsupportedToken {
+                token,
+            }));
+        }
+        Ok(self.flash_fee_amount.get())
+    }
+
+    // This function can reenter, but it doesn't pose a risk because it always
+    // preserves the property that the amount minted at the beginning is always
+    // recovered and burned at the end, or else the entire function will revert.
+    fn flash_loan(
+        &mut self,
+        receiver: Address,
+        token: Address,
+        amount: U256,
+        data: Bytes,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Self::Error> {
+        let max_loan = self.max_flash_loan(token, erc20);
+        if amount > max_loan {
+            return Err(Error::ExceededMaxLoan(ERC3156ExceededMaxLoan {
+                max_loan,
+            }));
+        }
+
+        let fee = self.flash_fee(token, amount)?;
+        if !Address::has_code(&receiver) {
+            return Err(Error::InvalidReceiver(ERC3156InvalidReceiver {
+                receiver,
+            }));
+        }
+        erc20._mint(receiver, amount)?;
+        let loan_receiver = IERC3156FlashBorrower::new(receiver);
+        let loan_return = loan_receiver
+            .on_flash_loan(
+                Call::new_in(self),
+                msg::sender(),
+                token,
+                amount,
+                fee,
+                data.to_vec().into(),
+            )
+            .map_err(|_| {
+                Error::InvalidReceiver(ERC3156InvalidReceiver { receiver })
+            })?;
+        if loan_return != BORROWER_CALLBACK_VALUE {
+            return Err(Error::InvalidReceiver(ERC3156InvalidReceiver {
+                receiver,
+            }));
+        }
+
+        let allowance = amount
+            .checked_add(fee)
+            .expect("allowance should not exceed `U256::MAX`");
+        erc20._spend_allowance(receiver, contract::address(), allowance)?;
+
+        let flash_fee_receiver = self.flash_fee_receiver_address.get();
+
+        if fee.is_zero() || flash_fee_receiver.is_zero() {
+            erc20._burn(receiver, allowance)?;
+        } else {
+            erc20._burn(receiver, amount)?;
+            erc20._transfer(receiver, flash_fee_receiver, fee)?;
+        }
+
+        Ok(true)
+    }
+}
+
+// TODO: unignore all tests once it's possible to mock contract address.
+// NOTE: double check that the tests assert the correct and expected things.
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use alloy_primitives::{address, uint, Address, U256};
+    use stylus_sdk::msg;
+
+    use super::{Erc20, Erc20FlashMint, Error, IErc3156FlashLender};
+
+    const ALICE: Address = address!("A11CEacF9aa32246d767FCCD72e02d6bCbcC375d");
+    const TOKEN_ADDRESS: Address =
+        address!("dce82b5f92c98f27f116f70491a487effdb6a2a9");
+    const INVALID_TOKEN_ADDRESS: Address =
+        address!("dce82b5f92c98f27f116f70491a487effdb6a2aa");
+
+    #[motsu::test]
+    #[ignore]
```

Comment:
I see that we can check these test cases with approach similar to [Erc721 Metadata](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/deb20ecff90bc2473e5598f90acb7eb89fbbe2d9/contracts/src/token/erc721/extensions/uri_storage.rs#L128). Declare a contract that contains `Erc20` and `Erc20FlashMint` inside. Similar to what you have in [FlashMintExample](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/deb20ecff90bc2473e5598f90acb7eb89fbbe2d9/examples/erc20-flash-mint/src/lib.rs#L18).

Then test cases should work

Reply:
Otherwise we cannot use to contract's storages in motsu test case

Reply:
Since FlashMint relies upon `contract::address()` make sense to wait a bit, when motsu can support it

Reply:
Fair enough but we should add a corresponding GH issue to update these tests when it will be possible.

Reply:
https://github.com/OpenZeppelin/rust-contracts-stylus/issues/470